Fix for NUTCH-2296: Elasticsearch Indexing Over Rest

bmzhao opened the pull request:
Open Elasticsearch to the option of REST-based indexing, via another indexing plugin implemented using Jest, allowing the use of https.

bmzhao commented:
Hello Lewis,

Here was that plugin I worked on a few months ago; unfortunately, I was running into methodNotFound errors using the Jest dependency. It turned out to be a versioning conflict between several apache http libraries (httpclient, httpcore, httpcore-nio). I tried adding them to the plugin's ivy.xml file without any success. However, after I added them to the root ivy.xml file, I'm no longer seeing these errors. 

Is there a better way around this issue besides modifying the root ivy.xml file? And if I resolve it, is there any chance this PR would get merged?

Thanks!

lewismc commented:
Hi @bmzhao I know EXACTLY what you are referring to and this dependency management issue is something we've been aware of for a while.
I'll review your current PR and come back to you on that point in due course. Thanks for your contribution this is great.

lewismc commented on src/plugin/indexer-elastic-rest/howto_upgrade_es.txt:
```diff
@@ -0,0 +1,6 @@
+1. Upgrade elasticsearch dependency in src/plugin/indexer-elastic/ivy.xml
```
Can you include trivial license header here please.

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/ElasticRestIndexWriter.java:
```diff
@@ -0,0 +1,336 @@
+ * limitations under the License.
+ */
+
+//TODO trust self signed and non matching certs: http://stackoverflow.com/questions/2893819/telling-java-to-accept-self-signed-ssl-certificate
```
Can you log issues in Jira for these 2 TODO's and also link to the original https://issues.apache.org/jira/browse/NUTCH-2296 Thanks

lewismc commented:
Can you please refactor all of your code inline withe eclipse formatter which can be found at
https://github.com/apache/nutch/blob/master/eclipse-codeformat.xml
if you are not using Elcipse, then essentially it is 2 line indents. Thank you @bmzhao

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/ElasticRestIndexWriter.java:
```diff
@@ -0,0 +1,336 @@
+                }
+            }).build();
+        } catch (NoSuchAlgorithmException | KeyManagementException | KeyStoreException e) {
+            e.printStackTrace();
```
Can you actually LOG the stacktrace instead of printing it?
LOG.error("Error detected whilst... blah blah blah", e).
Thank you

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/ElasticRestIndexWriter.java:
```diff
@@ -0,0 +1,336 @@
+            }
+            jestClientFactory.setHttpClientConfig(builder.build());
+        } else {
+            throw new RuntimeException("No host and port specified");
```
Is there a more specific exception here over and above merely RuntimeException? It would be nice to LOG and capture it such that it is maintained within application logs. Thanks

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/ElasticRestIndexWriter.java:
```diff
@@ -0,0 +1,336 @@
+        bulkDocs++;
+
+        if (bulkDocs >= maxBulkDocs || bulkLength >= maxBulkLength) {
+            LOG.info("Processing bulk request [docs = " + bulkDocs + ", length = "
```
Can you please use parametertized logging here e.g. 
```LOG.info("Processing bulk request [docs = {},...", bulkDocs, bulkLength, indexedDocs)

Please see http://slf4j.org/faq.html#logging_performance

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/ElasticRestIndexWriter.java:
```diff
@@ -0,0 +1,336 @@
+                    throw new RuntimeException();
+                }
+                long msWaited = System.currentTimeMillis() - beforeWait;
+                LOG.info("Previous took in ms " + millis
```
Same here, please use parameterized logging

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/ElasticRestIndexWriter.java:
```diff
@@ -0,0 +1,336 @@
+                LOG.info("Previous took in ms " + millis
+                        + ", including wait " + msWaited);
+            } catch (InterruptedException | ExecutionException e) {
+                LOG.error("Error waiting for result ", e);
```
This is much better. Thanks for logging exception as well.

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/ElasticRestIndexWriter.java:
```diff
@@ -0,0 +1,336 @@
+    @Override
+    public void close() throws IOException {
+        // Flush pending requests
+        LOG.info("Processing remaining requests [docs = " + bulkDocs
```
Patametrrized logging please

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/ElasticRestIndexWriter.java:
```diff
@@ -0,0 +1,336 @@
+        if (StringUtils.isBlank(host)) {
+            String message = "Missing elastic.host. It should be set in nutch-site.xml ";
+            message += "\n" + describe();
+            LOG.error(message);
```
Please sort this out.

lewismc commented on src/plugin/indexer-elastic-rest/src/java/org/apache/nutch/indexwriter/elasticrest/package-info.java:
```diff
@@ -0,0 +1,22 @@
+ */
+
+/**
+ * Index writer plugin for <a href="http://www.elasticsearch.org/">Elasticsearch</a>.
```
REST-based

lewismc commented:
Does this plugin have authentication built in?

bmzhao commented:
Hi @lewismc, I've refactored my branch to try to resolve all the issues: 
1. I've added the Apache License as comments to the howto_upgrade_es.txt
2. I've installed an Eclipse Formatting Plugin into Intellij, and have run it on the plugin; everything is now 2 spaces indented
3. I've refactored all 'e.printstacktrace()' calls to be Logger.error() calls instead (sorry for the inconsistency, my IDE autogenerates these for try catch blocks, and I forgot to replace some of them with Logger.error() calls)
4. I've changed all Logger method calls to use parameterized logging
5. I've also changed the RuntimeException to be an IllegalStateException instead, since the host and port information must be provided (and it would be an illegal state to proceed without them)
6. I've also fixed the package.info to mention the RESTful nature of the plugin
7. I'm still sort of confused on what the setConf method was supposed to do, but I've refactored it to mimic the other elastic index writer plugin; I'm not sure how the Conf object passed in this method differs from the JobConf object passed in the "open" method, so I have the open method handling the class member initialization (since that's what the other plugins seem to do). 

The plugin also does support authentication! If you've enabled the https option in nutch-site.xml, and the username and password fields are both not empty, the plugin will attempt to use them to perform a basic-auth request. I've also deleted one of the todo's since I've included the code to perform NoOp verification on hostnames (if the option is set to true in nutch-site) as well as trusting self-signed certs.

Thanks for reviewing my pull request!
Please let me know if there are any other changes I should fix; I'll be happy to commit them the next weekend I'm free.

lewismc commented:
Does anyone have an objection to merging this PR into master? Thank you for the patch and explanation above @bmzhao

purbon commented:
This contribution makes a lot of sense, moving out of the transport protocol makes a lot of sense. however I find missing why not also update to a newer version? being on 1.4.1 is a very old one.

purbon commented:
I amend myself the 1.4 is in the old 1.x branch, not in the new one. however also good that the code might be updated to latest versions so for example can benefit from the java rest client.

lewismc commented:
@purbon I think we can make the upgrades in a separate pull request.
